dataset/test/assign_test.cpp

```diff
@@ -0,0 +1,60 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2021 Scipp contributors (https://github.com/scipp)
+#include <gtest/gtest.h>
+
+#include "scipp/dataset/dataset.h"
+#include "scipp/dataset/except.h"
+
+using namespace scipp;
+using namespace scipp::dataset;
+
+struct AssignTest : public ::testing::Test {
+  AssignTest() {}
+
+protected:
+  Dimensions dims{Dim::X, 3};
+  Variable data = makeVariable<double>(dims, Values{1, 2, 3});
+  Variable x = makeVariable<double>(dims, Values{1, 1, 3});
+  Variable mask = makeVariable<bool>(dims, Values{true, false, true});
+  DataArray array{data, {{Dim::X, copy(x)}}, {{"mask", mask}}};
+};
+
+TEST_F(AssignTest, self) {
+  const auto original = copy(array);
+  EXPECT_EQ(array.assign(array), original);
+}
+
+TEST_F(AssignTest, coord_fail) {
+  const auto original = copy(array);
+  EXPECT_THROW(array.assign(array.slice({Dim::X, 0, 1})),
+               except::CoordMismatchError);
+  EXPECT_EQ(array, original);
+  EXPECT_THROW(array.slice({Dim::X, 0, 1}).assign(array.slice({Dim::X, 2, 3})),
+               except::CoordMismatchError);
+  EXPECT_EQ(array, original);
+}
+
+TEST_F(AssignTest, mask_propagation) {
+  const auto original = copy(array);
+  // Mask values get copied
+  array.slice({Dim::X, 0}).assign(original.slice({Dim::X, 1}));
+  EXPECT_EQ(array.masks()["mask"],
+            makeVariable<bool>(dims, Values{false, false, true}));
+  array.slice({Dim::X, 0}).assign(original.slice({Dim::X, 2}));
+  EXPECT_EQ(array.masks()["mask"],
+            makeVariable<bool>(dims, Values{true, false, true}));
+  // Mask not in source is preserved unchanged
+  array.masks().set("other", mask);
+  array.slice({Dim::X, 0}).assign(original.slice({Dim::X, 1}));
+  EXPECT_EQ(array.masks()["mask"], mask);
+  // Extra mask is added
+  auto extra_mask = copy(array);
+  extra_mask.masks().set("extra", mask);
+  array.assign(extra_mask.slice({Dim::X, 1}));
+  EXPECT_TRUE(array.masks().contains("extra"));
+  // Extra masks added to mask dict of slice => silently dropped
+  extra_mask.masks().set("dropped", mask);
+  EXPECT_NO_THROW(
+      array.slice({Dim::X, 0}).assign(extra_mask.slice({Dim::X, 1})));
+  EXPECT_FALSE(array.masks().contains("dropped"));
```

Comment:
@OwenArnold @jl-wynen This is the one remaining weird bit: This mask gets silently dropped. This is kind of the same reason as coord being silently "dropped" in the following:
```python
ds['a'].coords['x'] = x
```
Is there a sane way to avoid this?

Reply:
The mask could be extended to cover the entire array but be `false` everywhere outside of the slice. This would correctly mask out any elements in the RHS and play nicely with a later assignments that also contain the mask.
But it would implicitly allocate a potentially large amount of memory.

Reply:
I think that would make sense if we think of this in terms of `__setitem__`:
```python
da['x', 2:4] = masked
```
The implementation of `DataArray.__setitem__` could insert the new masks. The fact that the current implementaton in C++ using `assign` on the slice doesn't work is kind of an implementation artifact.

However, what about chained slices:

```python
da['x', 2:4]['y', 2:4] = masked
```
This would call `__setitem__` on `da['x', 2:4]`, i.e., would insert into a slice of the mask dict and would again NOT at the mask to `da`. Is the implication that we need a syntax that allows slicing all dims in one go, rather than using chaining?

Reply:
I am in favour of such a syntax anyway. But we still need to fix the chained case. If it is not possible or too complicated to check if the mask already exists in `da` or to expand it to the required size and assign it, then we should raise an exception. I think it is better to prohibit this operation than to store masked values and drop the mask.

Reply:
Having thought about this some more, I have reached the intermediate conclusion that the only way to prevent unwanted masks modifications is to a `readonly` flag to `Variable`. I suggest discussing this in a call.